x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items_details.ts

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ProcessorEvent } from '@kbn/observability-plugin/common/processor_event';
+import {
+  kqlQuery,
+  rangeQuery,
+  termQuery,
+} from '@kbn/observability-plugin/server/utils/queries';
+import {
+  AGENT_NAME,
+  AGENT_VERSION,
+  SERVICE_ENVIRONMENT,
+  SERVICE_LANGUAGE_NAME,
+  SERVICE_NAME,
+  SERVICE_NODE_NAME,
+} from '../../../common/elasticsearch_fieldnames';
+import { environmentQuery } from '../../../common/utils/environment_query';
+import { AgentName } from '../../../typings/es_schemas/ui/fields/agent';
+import { APMEventClient } from '../../lib/helpers/create_es_client/create_apm_event_client';
+import { RandomSampler } from '../../lib/helpers/get_random_sampler';
+
+interface AggregationParams {
+  environment: string;
+  serviceName?: string;
+  agentLanguage?: string;
+  kuery: string;
+  apmEventClient: APMEventClient;
+  maxNumServices: number;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}
+
+export async function getAgentsDetails({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  apmEventClient,
+  maxNumServices,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const response = await apmEventClient.search('get_agent_details', {
+    apm: {
+      events: [ProcessorEvent.metric],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              exists: {
+                field: AGENT_NAME,
+              },
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              },
+            },
+            {
+              exists: {
+                field: SERVICE_NODE_NAME,
+              },
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...(serviceName ? termQuery(SERVICE_NAME, serviceName) : []),
+            ...(agentLanguage
+              ? termQuery(SERVICE_LANGUAGE_NAME, agentLanguage)
+              : []),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: maxNumServices,
+              },
+              aggs: {
+                serviceNodes: {
+                  terms: {
+                    field: SERVICE_NODE_NAME,
+                    size: maxNumServices,
+                  },
+                  aggs: {
```

Comment:
This can go up a level, to be a sub-agg of services

Reply:
Moving this one level up will be giving us the latest doc (agentName + agentVersion), being just one value per service. Instead we will like to go to agent instances, we can approx to agent instance placing the aggregation under `service.node.name`